ConnectionManager.php

```diff
@@ -0,0 +1,190 @@
+        $this->cacheManager->remember($tableName, $selectColumns, $conditions, $result);
+        return $result;
+    }
+
+
+    /**
+     * Generate an insert statement with bindings and pass it to database for execution
+     * @param $tableName
+     * @param $data
+     * @return bool
+     */
+    public function insert($tableName, $data)
+    {
+        $columns = implode(array_keys($data[0]), ',');
+        $query = "INSERT INTO `{$tableName}` ({$columns}) VALUES ";
+        $bindings = [];
+        foreach ($data as $row) {
+            $placeHolders = [];
+            foreach ($row as $columnName => $value) {
+                $placeHolderName = uniqid();
+                $placeHolders[] = ":{$placeHolderName}";
+                $bindings[$placeHolderName] = $value;
+            }
+            $placeHolders = implode(',', $placeHolders);
+            $query .= "({$placeHolders})";
+
+        }
+        return $this->db->execute($query, $bindings);
+    }
+
+    /**
+     * Generate an update statement with bindings and pass it to database for execution
+     * @param $tableName
+     * @param $data
+     * @param $conditions
+     * @return mixed
+     */
+    public function update($tableName, $data, $conditions)
+    {
+        $query = "UPDATE {$tableName} SET ";
+        $conditionsParsed = [];
+        $bindings = [];
+
+        foreach ($data as $column => $value) {
+            $placeHolderName = uniqid();
+            $query .= "{$column} = : {$placeHolderName} ";
+            $bindings[$placeHolderName] = $value;
+
+        }
+        foreach ($conditions as $condition) {
+            $placeHolderName = uniqid();
+
+
+            $conditionsParsed[] = " {$condition['column']} {$condition['operator']} :{$placeHolderName}";
+            $bindings[$placeHolderName] = $condition['value'];
+        }
+        $query .= "WHERE " . implode(' AND ', $conditionsParsed);
+
+        return $this->db->execute($query, $bindings);
+    }
+
+    /**
+     * Generate a delete statement with bindings and pass it to database for execution
+     * @param $tableName
+     * @param $conditions
+     * @return mixed
+     */
+    public function delete($tableName, $conditions)
+    {
+
+        $query = "DELETE   FROM {$tableName} WHERE ";
+        $conditionsParsed = [];
+        $bindings = [];
+
+        foreach ($conditions as $condition) {
+            $placeHolderName = uniqid();
+            $conditionsParsed[] = " {$condition['column']} {$condition['operator']} :{$placeHolderName}";
+            $bindings[$placeHolderName] = $condition['value'];
+        }
+        $query .= implode(' AND ', $conditionsParsed);
+
+        return $this->db->execute($query, $bindings);
+    }
+
+
+    /**
+     * Begins transaction if supported by db or throw exception
+     * @throws Exception
+     */
+    public function beginTransaction()
+    {
+        if (!$this->db->supportsTransactions()) {
+            $this->db->execute("BEGIN TRANSACTION;", []);
```

Comment:
@tsirigotis @gkalligeros This is and idiomatic statement and won't be correct with every DB that supports transactions.
e.g in MySQL the transaction statements are `START TRANSACTION, ROLLBACK and COMMIT` while in MSSQL is
`BEGIN, ROLLBACK, COMMIT`.

How can we implement this to be more flexible?

Reply:
That's correct ... We should let the database implementation implement start,commit and rollback transaction because this will be different for each dB implementation